src/StateService/Verification/VerificationContext.cs

```diff
@@ -0,0 +1,139 @@
+        private readonly Wallet wallet;
+        private readonly KeyPair keyPair;
+        private readonly int myIndex;
+        private readonly uint rootIndex;
+        private readonly ECPoint[] verifiers;
+        private int M => verifiers.Length - (verifiers.Length - 1) / 3;
+        private readonly Dictionary<int, byte[]> signatures = new Dictionary<int, byte[]>();
+
+        public int Retries;
+        public bool IsValidator => myIndex >= 0;
+        public int MyIndex => myIndex;
+        public ECPoint[] Verifiers => verifiers;
+        public ICancelable Timer;
+        public StateRoot StateRoot
+        {
+            get
+            {
+                if (root is null)
+                {
+                    using var snapshot = StateStore.Singleton.GetSnapshot();
+                    root = snapshot.GetStateRoot(rootIndex);
+                }
+                return root;
+            }
+        }
+        public ExtensiblePayload Message => payload;
+
+        public VerificationContext(Wallet wallet, uint index)
+        {
+            this.wallet = wallet;
+            Retries = 0;
+            myIndex = -1;
+            root = null;
+            rootIndex = index;
+            using SnapshotCache snapshot = Blockchain.Singleton.GetSnapshot();
+            verifiers = NativeContract.RoleManagement.GetDesignatedByRole(snapshot, Role.StateValidator, index);
+            if (wallet is null) return;
+            for (int i = 0; i < verifiers.Length; i++)
+            {
+                WalletAccount account = wallet.GetAccount(verifiers[i]);
+                if (account?.HasKey != true) continue;
+                myIndex = i;
+                keyPair = account.GetKey();
+                break;
+            }
+        }
+
+        public Vote CreateVote()
+        {
+            if (StateRoot is null) return null;
+            if (!signatures.TryGetValue(myIndex, out byte[] sig))
+            {
+                sig = StateRoot.Sign(keyPair);
+                signatures[myIndex] = sig;
+            }
+            return new Vote(rootIndex, myIndex, sig);
+        }
+
+        public bool AddSignature(int index, byte[] sig)
+        {
+            if (M <= signatures.Count) return false;
+            if (index < 0 || verifiers.Length <= index) return false;
+            if (signatures.ContainsKey(index)) return false;
+            Utility.Log(nameof(VerificationContext), LogLevel.Info, $"vote received, height={rootIndex}, index={index}");
+            ECPoint validator = verifiers[index];
+            byte[] hash_data = StateRoot?.GetHashData();
+            if (hash_data is null || !Crypto.VerifySignature(hash_data, sig, validator))
+            {
+                Utility.Log(nameof(VerificationContext), LogLevel.Info, "incorrect vote, invalid signature");
+                return false;
+            }
+            signatures.Add(index, sig);
+            return true;
+        }
+
+        private void CreateExtensiblePayload()
```

Comment:
Move this method to `CheckSignatures` in order to return false if sc it's not complete?

Reply:
Done. I remove the cath. Here we should throw exception beacuse it is not expected to happen. If it does, there must be  a bug.